### ISSUE TITLE
Can't make a clean Validator classes type hierarchy when having several grammars

### ISSUE THREAD

**Issue by vhemery** [eclipse/xsemantics#106]
Hello Lorenzo. I've been using Xsemantics very recently and it does a great job for what I expected. So first thing first, thank you very much !

(for the sake of readability I use ">" for extends "<" for extended by)

In my use case, I have not one, but 3 grammars that inherit from each other (GenericExpressions < StructuredDataExpressions < SpaceMCExpressions).
Each grammar has its own System, and it comes as a natural thing to extend the Systems in a same way that the grammars extend each other (GenericExpressionsSystem < StructuredDataExpressionsSystem < SpaceMCExpressionsSystem).

For the Generic grammar, I used the "validatorExtends" clause like advised in the documentation (http://xsemantics.sourceforge.net/xsemantics-documentation/XsemanticsSyntax.html#validatorExtends).
But for the other grammars, as I extend the Systems, I can't use the "validatorExtends" clause, and the new SystemValidator inherit directly from the extended grammar's SystemValidator. Hence I have to apply the trick from http://xsemantics.sourceforge.net/xsemantics-documentation/XsemanticsSyntax.html#withoutValidatorExtends and copy the getEPackages() implementation.

Finally, instead of having a clean Validator **Type** Hierarchy like :
`AbstractGenericExpressionsValidator`
`_<GenericExpressionsSystemValidator`
`___<GenericExpressionsValidator`
`_____<AbstractStructuredDataExpressionsValidator`
`_______<StructuredDataExpressionsSystemValidator`
`_________<StructuredDataExpressionsValidator`
`___________<AbstractSpaceMCExpressionsValidator`
`_____________<SpaceMCExpressionsSystemValidator`
`_______________<SpaceMCExpressionsValidator`
I end up with this **type** hierarchy :
`AbstractGenericExpressionsValidator`
`_<GenericExpressionsSystemValidator`
`___<GenericExpressionsValidator`
`_____<AbstractStructuredDataExpressionsValidator`
`___<StructuredDataExpressionsSystemValidator`
`______<SpaceMCExpressionsSystemValidator`
`_________<SpaceMCExpressionsValidator`
`______<StructuredDataExpressionsValidator`
`________<AbstractSpaceMCExpressionsValidator`

In my case, this does not cause too much trouble, since I do not have yet any relevant code in the Xtend Validator classes and everything is in the System. But in the general case, all code in StructuredDataExpressionsValidator or GenericExpressionsValidator never gets called when I use the SpaceMCExpressionsValidator.

Would it be possible for inheriting systems to allow the "validatorExtends" clause when specified with a sub-class of the super-system SystemValidator ? (sounds quite hard to check at runtime...)
I couldn't come to a better inheritance solution, as it would require double and lozange inheritance which is not an option with Java (e.g. StructuredDataExpressionsSystemValidator inheriting from AbstractStructuredDataExpressionsValidator and GenericExpressionsSystemValidator).
I have though about validator composition with "composedCheck" in the mwe2, but I couldn't find a good solution here either (causes validator redundancy).

I guess this multi-grammar+multi-system problematic hasn't been adressed yet in Xsemantic. Or have I missed something ?

Note that the currently implemented inheritance also makes sense in case we want only the type system but not the sub-grammar's validator, so I guess it's up to the system specifier in the end...

In case you want to discuss this feature request face to face, I'll be at EclipseConFrance next week.

**Comment by LorenzoBettini** [comment]
Hi

for sure Xsemantics is not meant for such an involved hierarchy... do you really need that level of grammar and system granularity?

I'm not sure what you propose with

> Would it be possible for inheriting systems to allow the "validatorExtends" clause when specified with a sub-class of the super-system SystemValidator ? (sounds quite hard to check at runtime...)

**Comment by vhemery** [comment]
I do need that grammar hierarchy and the possiblity to use them independantly. So I see no other way than having this system granularity.
Like I said, the reason this enhancement is not blocking for me is that I do not have grammar validation outside the system.
And even if I did, there would still be non-elegant workaround with code duplication or validation redundancy. So this feature is not a blocking one and it already works quite well in this use case.

> Would it be possible for inheriting systems to allow the "validatorExtends" clause when specified with a sub-class of the super-system SystemValidator ? (sounds quite hard to check at runtime...)
I am trying to get closer to the standard validators class inheritance I have with 1 System or 1 grammar. When I extend a super-system, the new SystemValidator class inherits directly from the super-system SystemValidator class which causes this messy class hierarchy I mentioned.
(by "package hierarchy" I meant "type hierarchy", sorry for the typo)

A straigh-line inheritance hierarchy would be better. Hence, I think my SystemValidator should better (correct me if I'm wrong here) inherit from the abstract grammar validator, which itself inherit indirectly from the super-system SystemValidator. So I would like to use the "validatorExtends" clause, just like I did in the parent system. I think this should be possible to let user define it as long as they respect the constraint that the SystemValidator must inherit (indirectly) from the super-system SystemValidator. On the other hand, I'm not sure how this constraint can be checked by xsemantics...
